clouddriver-ecs/src/integration/java/com/netflix/spinnaker/clouddriver/ecs/test/CreateServerGroupForExistingServiceSpec.java

```diff
@@ -0,0 +1,323 @@
+
+import static io.restassured.RestAssured.get;
+import static io.restassured.RestAssured.given;
+import static org.hamcrest.Matchers.containsString;
+import static org.hamcrest.Matchers.notNullValue;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.mockito.ArgumentMatchers.*;
+import static org.mockito.Mockito.*;
+
+import com.amazonaws.services.applicationautoscaling.AWSApplicationAutoScalingClient;
+import com.amazonaws.services.applicationautoscaling.model.DescribeScalableTargetsRequest;
+import com.amazonaws.services.applicationautoscaling.model.DescribeScalableTargetsResult;
+import com.amazonaws.services.ecs.AmazonECS;
+import com.amazonaws.services.ecs.model.*;
+import com.amazonaws.services.elasticloadbalancingv2.AmazonElasticLoadBalancing;
+import com.amazonaws.services.elasticloadbalancingv2.model.DescribeTargetGroupsRequest;
+import com.amazonaws.services.elasticloadbalancingv2.model.DescribeTargetGroupsResult;
+import com.amazonaws.services.elasticloadbalancingv2.model.TargetGroup;
+import com.netflix.spinnaker.clouddriver.artifacts.ArtifactCredentialsRepository;
+import com.netflix.spinnaker.clouddriver.artifacts.ArtifactDownloader;
+import com.netflix.spinnaker.clouddriver.artifacts.config.ArtifactCredentials;
+import com.netflix.spinnaker.clouddriver.aws.security.NetflixAmazonCredentials;
+import com.netflix.spinnaker.clouddriver.ecs.EcsSpec;
+import com.netflix.spinnaker.kork.artifacts.model.Artifact;
+import io.restassured.http.ContentType;
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.util.Collections;
+import java.util.Date;
+import java.util.List;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.DisplayName;
+import org.junit.jupiter.api.Test;
+import org.mockito.ArgumentCaptor;
+import org.mockito.stubbing.Answer;
+import org.springframework.boot.test.mock.mockito.MockBean;
+
+public class CreateServerGroupForExistingServiceSpec extends EcsSpec {
+
+  @MockBean ArtifactDownloader mockArtifactDownloader;
+
+  @MockBean ArtifactCredentialsRepository mockArtifactCredentialsRepository;
+
+  private ArtifactCredentials mockArtifactCredentials = mock(ArtifactCredentials.class);
+
+  private AWSApplicationAutoScalingClient mockAWSApplicationAutoScalingClient =
+      mock(AWSApplicationAutoScalingClient.class);
+
+  private AmazonECS mockECS = mock(AmazonECS.class);
+
+  private AmazonElasticLoadBalancing mockELB = mock(AmazonElasticLoadBalancing.class);
+
+  @BeforeEach
+  public void setup() {
+
+    // mocking calls
+    when(mockECS.describeServices(any(DescribeServicesRequest.class)))
+        .thenReturn(
+            new DescribeServicesResult()
+                .withServices(
+                    Collections.singletonList(
+                        new Service()
+                            .withServiceName("createService")
+                            .withCreatedAt(new Date())
+                            .withStatus("INACTIVE"))));
```

Comment:
does it have to be `INACTIVE`? what happens if it's running?

Reply:
Good question @allisaurus 
So we are mocking the call here and setting the status to INACTIVE in order to get the service version as 2. If we keep this status as ACTIVE or DRAINING then [this particular code](https://github.com/paragbhingre/clouddriver/blob/f86b76ae24ef05ad100ccac372b4051164cfceb8/clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/deploy/EcsServerGroupNameResolver.java#L95-L115) would run max 5 times and nextServerGroupName would become v005 or v006. So in order to keep it simple we considered that the existing service is in INACTIVE state in order to get the nextServerGroupName in terms of v002.
Does that make sense?

Reply:
thanks for the code link! it looks like that logic will [actually do the same thing](https://github.com/paragbhingre/clouddriver/blob/f86b76ae24ef05ad100ccac372b4051164cfceb8/clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/deploy/EcsServerGroupNameResolver.java#L105-L106) if this describe call returns empty (vs. with one inactive service). What do you think about doing that instead, since it's simpler?

Reply:
Hey @allisaurus ,
If I try to return empty describe call (i.e. withServices but without any other properties) as you recommend then [taken slots code](https://github.com/paragbhingre/clouddriver/blob/611940b7676ee5b43f3066dfd78bc08796d01e8c/clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/deploy/EcsServerGroupNameResolver.java#L71-L82) successfully returns a slot for server group with the next sequence i.e. -v001. 
But then [this particular code](https://github.com/paragbhingre/clouddriver/blob/f86b76ae24ef05ad100ccac372b4051164cfceb8/clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/deploy/EcsServerGroupNameResolver.java#L95-L115) fails to consider service as empty and gives NPE.

Reply:
Hello @allisaurus , may I know your thoughts on this one so that I will push the remaining changes and we are good to go with this test?

Reply:
@paragbhingre which part of the code is returning the NPE? [Line 105](https://github.com/paragbhingre/clouddriver/blob/f86b76ae24ef05ad100ccac372b4051164cfceb8/clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/deploy/EcsServerGroupNameResolver.java#L105)?

Reply:
@allisaurus  `result.getServices().get(0).getStatus().equals("INACTIVE")` is throwing NPE as service is not empty so it tries to check the status property and we are not setting it's status property.

Reply:
oh! so it's not recognizing the service list as empty - but aren't we just returning an empty `DescribeServicesRequest` ? That was the [suggestion](https://github.com/spinnaker/clouddriver/pull/5073#discussion_r526356346)

Reply:
OK so I think I understand what's happening (finally). First time we call **DescribeServices** it needs to return something to indicate that we need to increment the server group version. But the 2nd time it's called it needs to return empty (or `INACTIVE`) to indicate that the `v001` "slot" is free to use.

What you have allows the same mock response to be returned both times and results in the expected SG name. I am OK with this! Though it is, realistically, a less likely scenario than the first one existing & the second one not. If you are inclined to try and mock the 2 sequential responses go ahead, though I will not continue to block on this in order to wrap up this PR.

Reply:
So after final trial as per your suggestion on this, the outcome is, even if we try to mock 2 sequential responses, code doesn't know which one to pick 1st and that's why the problem remains. Moreover, as per your suggestion when we send empty response in **DescribeServices** then [getTokenSlots code](https://github.com/spinnaker/clouddriver/blob/611940b7676ee5b43f3066dfd78bc08796d01e8c/clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/deploy/EcsServerGroupNameResolver.java#L75-L78) fails to find a service and does not create a slot for it which means in return we get the service with v000 and not v001. 
So I guess we need to keep this INACTIVE condition as is to pass this particular test case.